core/test/matrix/fbcsr_sample.hpp

```diff
@@ -0,0 +1,244 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2020, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+#ifndef GKO_CORE_MATRIX_TEST_FBCSR_SAMPLE_HPP
+#define GKO_CORE_MATRIX_TEST_FBCSR_SAMPLE_HPP
+
+#include <ginkgo/core/base/matrix_data.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/matrix/diagonal.hpp>
+#include <ginkgo/core/matrix/fbcsr.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+namespace gko {
+namespace testing {
+
+
+/// Generates the same sample block CSR matrix in different formats
+/** This currently a 6 x 12 matrix with 3x3 blocks.
+ * Assumes that the layout within each block is row-major.
+ * Generates complex data when instantiated with a complex value type.
+ */
+template <typename ValueType, typename IndexType>
+class FbcsrSample {
```

Comment:
The code in `fbcsr_sample.cpp` is already being used in two different test executables. It might be used in more in the future, though I'm not sure. Do you still prefer the implementation in the hpp? I would normally create a static library for this kind of thing; we can do that if it's better. Looks like it's creating a .so by default.

Reply:
You can also include a header in the places where you need them, for example this is what we do with `core/test/utils/matrix_generator.hpp` and others, which are pretty much included in every test right now.

Reply:
Yeah but that would lead to compiling the same code multiple times for no good reason. What's the disadvantage of having a (static or dynamic) library for this? If you still think it's better header-only, I can do that, no problem. This particular file may not add much to the compile time, but I wonder if that's a policy that should be adopted.

Reply:
I think it would be much better header only for the fact we always do this, but also because the current way of generating a library breaks the new algorithm script, and the compilation is non standard and is another place we need to be careful if we do anything (change Ginkgo flags, use different C++ standard, ...).

Much more importantly, I think this class should actually be removed altogether, or at least shrinked to a bare minimum. A lot of functionality I think are really duplicate from our other ones (standard conversions) etc, which we test separately and know work (dense->csr for example, or `read()`).

What could be done instead would be to simply have a base `Dense` matrix or something else which would be compared against in order to remove most of the complexity from this file. That would mean that to test Fbcsr you would need only one thing: ensure that dense->fbcsr (and reverse) conversion work properly, in a manual fashion. Afterwards, you can call the transpose of dense (or whichever other functionality), and compare against fbcsr's equivalent functionality with a conversion.

The rationale for this is that for now the values are duplicated everywhere which makes it really hard to edit the values/layout of this `FbcsrSample` as it requires editing pretty much every function in the related `.cpp` file.